src/utils/sample/discrete_gauss.rs

```diff
@@ -0,0 +1,252 @@
+// Copyright © 2023 Niklas Siemer
+//
+// This file is part of qFALL-math.
+//
+// qFALL-math is free software: you can redistribute it and/or modify it under
+// the terms of the Mozilla Public License Version 2.0 as published by the
+// Mozilla Foundation. See <https://mozilla.org/en-US/MPL/2.0/>.
+
+//! This module includes core functionality to sample according to the
+//! discrete gaussian distribution.
+
+use super::uniform::{get_rng, sample_uniform_rejection};
+use crate::{error::MathError, integer::Z, rational::Q, traits::Pow};
+use rand::RngCore;
+
+#[allow(dead_code)]
+/// Chooses a sample according to the discrete Gaussian distribution out of
+/// `[center - ⌈s * log_2(n)⌉ , center + ⌊s * log_2(n)⌋ ]`.
+///
+/// Parameters:
+/// - `n`: specifies the range from which is sampled
+/// - `center`: specifies the position of the center with peak probability
+/// - `s`: specifies the Gaussian parameter, which is proportional
+/// to the standard deviation `sigma * sqrt(2 * pi) = s`
+///
+/// Returns a sample chosen according to the specified discrete Gaussian distribution or
+/// a [`MathError`] if the specified parameters were not chosen appropriately,
+/// i.e. `n > 1` and `s > 0`.
+///
+/// # Examples
+/// ```compile_fail
+/// use qfall_math::{integer::Z, rational::Q};
+/// use qfall_math::utils::sample::discrete_gauss::sample_z;
+/// let n = Z::from(1024);
+/// let center = Q::ZERO;
+/// let gaussian_parameter = Q::ONE;
+///
+/// let sample = sample_z(&n, &center, &gaussian_parameter).unwrap();
+/// ```
+///
+/// # Errors and Failures
+/// - Returns a [`MathError`] of type [`InvalidIntegerInput`](MathError::InvalidIntegerInput)
+/// if the `n <= 1` or `s <= 0`.
+pub(crate) fn sample_z(n: &Z, center: &Q, s: &Q) -> Result<Z, MathError> {
+    // TODO: Change this functions signature to use std_deviation/ sigma and not Gaussian parameter
+    if n <= &Z::ONE {
+        return Err(MathError::InvalidIntegerInput(format!(
+            "The value {} was provided for parameter n of the function sample_z.
+            This function expects this input to be bigger than 1.",
+            n
+        )));
+    }
+    if s <= &Q::ZERO {
+        return Err(MathError::InvalidIntegerInput(format!(
+            "The value {} was provided for parameter s of the function sample_z.
+            This function expects this input to be bigger than 0.",
+            s
+        )));
+    }
+
+    let lower_bound = (center - s * n.log(&2).unwrap()).ceil();
+    let upper_bound = (center + s * n.log(&2).unwrap()).floor();
+    let interval_size = &upper_bound - &lower_bound;
+
+    // sample x in [c - s * log_2(n), c + s * log_2(n)]
+    let mut sample = &lower_bound + sample_uniform_rejection(&interval_size).unwrap();
+
+    // rejection sample
+    let rng = get_rng();
+    // TODO: Change to Q::from_f64 once it works appropriately for large scale
+    while gaussian_function(&sample, center, s) < Q::try_from((&rng.next_u64(), &u64::MAX)).unwrap()
```

Comment:
```suggestion
    while gaussian_function(&sample, center, s) <= Q::try_from((&rng.next_u64(), &u64::MAX)).unwrap()
```

I maybe did something wrong in the challenge:
I have looked back into https://eprint.iacr.org/2007/432.pdf page 14. 4.1: 'Then with probability
ρs(x − c) ∈ (0, 1], output x, otherwise repeat.'
Assuming rng samples from {0,1}^*, pls correct me if im wrong.

For probability 0.5 (binary 0.1) the sample should be accepted with probability 0.5 and rejected with probability 0.5. But 0.1 is never lower  than 0.1 or 0.0, it would be accepted with prob. 1